src/main/java/org/jenkinsci/plugins/workflow/actions/ArgumentsAction.java

```diff
@@ -0,0 +1,169 @@
+/*
+ * The MIT License
+ *
+ * Copyright (c) 2017, CloudBees, Inc.
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+package org.jenkinsci.plugins.workflow.actions;
+
+import org.jenkinsci.plugins.workflow.graph.FlowNode;
+import org.jenkinsci.plugins.workflow.graph.StepNode;
+import org.jenkinsci.plugins.workflow.steps.Step;
+import org.jenkinsci.plugins.workflow.steps.StepDescriptor;
+
+import javax.annotation.CheckForNull;
+import javax.annotation.Nonnull;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Map;
+
+/**
+ * Stores some or all of the arguments used to create and configure the {@link Step} executed by a {@link FlowNode}.
+ * This allows you to inspect information supplied in the pipeline script and otherwise discarded at runtime.
+ * Supplied argument values can be hidden and replaced with a {@link NotStoredReason} for security or performance.
+ */
+public abstract class ArgumentsAction implements PersistentAction {
+
+    /** Used as a placeholder for arguments not stored for various reasons */
+    public enum NotStoredReason {
+        /** Denotes an unsafe value that cannot be stored/displayed due to sensitive info */
+        MASKED_VALUE,
+
+        /** Denotes an object that is too big to retain, such as strings exceeding {@link #MAX_STRING_LENGTH} */
+        OVERSIZE_VALUE
+    }
+
+    /** Largest string we'll persist in the Step, for performance reasons */
+    public static final int MAX_STRING_LENGTH = 1024;
```

Comment:
@stephenc Agree that the length naming is not as helpful as it might be -- but there's no sense in throwing away perfectly useful arguments just because one is oversized (such as file content). 

Would you and @jglick be satisfied if I renamed it to "MAX_ARRAY_LENGTH" and moved the workflow-cps filtering logic here in some fashion? (Assuming that will work, I'll have to double check we don't need anything provided by workflow-cps).

I've also considered offering an API for StepDescriptors that want to provide their own filtering/transformation of stored arguments beyond the security filtering (as mentioned to Vivek above).  But we just don't have a need for it yet, so YAGNI for now...